### PR TITLE
Dynamic auth generation

### PR DESCRIPTION
**Pull request by tupe12334**

### CODE REVIEW

File: packages/amplication-data-service-generator/src/server/auth/createAuth.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import { AppInfo, Module } from "../../types";
+import { SRC_DIRECTORY } from "../constants";
+import { createDefaultGuard } from "./guards/createDefaultGuard";
+
+export const authPath = `${SRC_DIRECTORY}/auth`;
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/auth/createAuth.ts:5]: move to ./constants.ts

File: packages/amplication-data-service-generator/src/tests/__snapshots__/create-data-service.spec.ts.snap [hunk 2]
```diff
@@ -2777,7 +2777,7 @@ export class BasicStrategy extends PassportStrategy(Strategy) {
 
 export class BasicAuthGuard extends AuthGuard(\\"basic\\") {}
 ",
-  "server/src/auth/defaultAuth.guard.ts": "import { BasicAuthGuard } from \\"./basicAuth.guard\\";
+  "server/src/auth/defaultAuth.guard.ts": "import { BasicAuthGuard } from \\"./BasicAuthGuard\\";
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/tests/__snapshots__/create-data-service.spec.ts.snap:2780]: this should not change

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-data-service-generator/src/server/auth/guards/createDefaultGuard.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+): AuthGuardMetaData {
+  const data: AuthGuardMetaData = { fileName: "", path: "", className: "" };
+  switch (setAuthGuard) {
+    case EnumAuthProviderType.Http:
+      data.className = "BasicAuthGuard";
+      data.fileName = "basicAuth";
+      data.path = `${AUTH_PATH}/${data.fileName}.guard.ts`;
+      break;
+    case EnumAuthProviderType.Jwt:
+      data.fileName = "jwt-auth";
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/auth/guards/createDefaultGuard.ts:53]: keep the same convention like basicAuth
```suggestion
      data.fileName = "jwtAuth";
```

File: packages/amplication-data-service-generator/src/server/auth/guards/createDefaultGuard.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+}
+function getMetaDataForAuthGuard(
+  setAuthGuard: EnumAuthProviderType
+): AuthGuardMetaData {
+  const data: AuthGuardMetaData = { fileName: "", path: "", className: "" };
+  switch (setAuthGuard) {
+    case EnumAuthProviderType.Http:
+      data.className = "BasicAuthGuard";
+      data.fileName = "basicAuth";
+      data.path = `${AUTH_PATH}/${data.fileName}.guard.ts`;
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/auth/guards/createDefaultGuard.ts:50]: keep same structure for both providers
```suggestion
      data.path = `${AUTH_PATH}/basic/${data.fileName}.guard.ts`;
```

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
Great work!
Please see some minor changes

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
in the "basic" folder we use "guard" folder while in the "jwt" folder we use "guards" folder (in plural).
Instead, I suggest we remove the folder entirely and put the guards in the same folder like the strategy

File: packages/amplication-data-service-generator/src/server/static/src/auth/gqlDefaultAuth.guard.ts [hunk 5]
```diff
@@ -1,3 +1,3 @@
-import { GqlBasicAuthGuard } from "./gqlBasicAuth.guard";
+import { GqlBasicAuthGuard } from "./basic/gqlBasicAuth.guard";
 
 export class GqlDefaultAuthGuard extends GqlBasicAuthGuard {}
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/static/src/auth/gqlDefaultAuth.guard.ts:3]: This should be dynamic

File: packages/amplication-data-service-generator/src/server/static/src/auth/jwt/jwtAuth.guard.ts [hunk 6]
```diff
@@ -1,5 +1,3 @@
-import { Injectable } from "@nestjs/common";
 import { AuthGuard } from "@nestjs/passport";
 
-@Injectable()
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/static/src/auth/jwt/jwtAuth.guard.ts:4]: This should be consistent between jwtAuth and basicAuth

**Review by yuval-hazaz — APPROVED** [review]
There are two comments, to make things a little bit easier, please fix these two new comments in a separate PR with rebase to this PR

File: packages/amplication-data-service-generator/src/server/static/src/auth/gqlDefaultAuth.guard.ts (hunk 5 above)
Comment (tupe12334) [reply packages/amplication-data-service-generator/src/server/static/src/auth/gqlDefaultAuth.guard.ts:3]: what is your proposal? GqlDefaultAuthGuard made dynamically or extends the defaultAuthGuard?

Comment (yuval-hazaz) [reply packages/amplication-data-service-generator/src/server/static/src/auth/gqlDefaultAuth.guard.ts:3]: Like we discussed yesterday, it is best to use a single guard for GQL that extended defaultGuard that adds the "getRequest" function.
This way we only need to add a single guard file per provider, and all guards will be extended for GraphQL via the default guard.

File: packages/amplication-data-service-generator/src/server/create-dotenv.template.env [hunk 7]
```diff
@@ -5,4 +5,5 @@ POSTGRESQL_URL=postgres://${dbUser}:${dbPassword}@${dbHost}:${dbPort}${dbName}
 BCRYPT_SALT=10
 COMPOSE_PROJECT_NAME=amp_${appId}
 JWT_SECRET_KEY=Change_ME!!!
-JWT_EXPIRATION=2d # more options in https://github.com/nestjs/jwt#jwtservicedecodetoken-string-options-decodeoptions-object--string
\ No newline at end of file
+# more options in https://github.com/nestjs/jwt#jwtservicedecodetoken-string-options-decodeoptions-object--string
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/create-dotenv.template.env:8]: remove this link, all the details and instructions should be included in our docs site including links to external docs

File: packages/amplication-data-service-generator/src/server/resource/resolver/resolver.base.template.ts [hunk 8]
```diff
@@ -3,7 +3,7 @@ import * as graphql from "@nestjs/graphql";
 import * as apollo from "apollo-server-express";
 import * as nestAccessControl from "nest-access-control";
 // @ts-ignore
-import * as gqlDefaultAuthGuard from "../../auth/gqlDefaultAuth.guard";
+import { GqlDefaultAuthGuard } from "../../auth/gqlDefaultAuth.guard";
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/resource/resolver/resolver.base.template.ts:6]: For any future case, please do not include such changes with other changes. 
Any code refactoring especially ones that impact the code generation should be justified and included in a separate PR. 
for this case, just leave it as it is.

File: packages/amplication-data-service-generator/src/server/static/src/auth/auth.module.ts [hunk 9]
```diff
@@ -7,7 +7,7 @@ import { UserModule } from "../user/user.module";
 import { AuthController } from "./auth.controller";
 import { AuthResolver } from "./auth.resolver";
 import { AuthService } from "./auth.service";
-import { BasicStrategy } from "./basic.strategy";
+import { BasicStrategy } from "./basic/basic.strategy";
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/static/src/auth/auth.module.ts:10]: why do we need to import basicStrategy, and we do not need JwtStartegy?

If it is required, we probably need to add JwtStartegy. if not - please open a new issue and remove this line of code in a separate PR

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
Please include the requested changes in a separate PR/Branch rebased to this branch

File: packages/amplication-data-service-generator/src/server/static/src/auth/auth.module.ts (hunk 9 above)
Comment (tupe12334) [reply packages/amplication-data-service-generator/src/server/static/src/auth/auth.module.ts:10]: Look at the line below, we do import the JwtStrategy

**Pull request merged**